clients/drcachesim/tracer/physaddr.cpp

```diff
@@ -69,22 +69,37 @@ physaddr_t::physaddr_t()
     , count_(0)
 #endif
 {
-    // Nothing else.
-    // No destructor needed either: ifstream closes when destroyed.
+#ifdef LINUX
+    v2p_.table = nullptr;
+#endif
+}
+
+physaddr_t::~physaddr_t()
+{
+#ifdef LINUX
+    if (v2p_.table != nullptr)
+        hashtable_delete(&v2p_);
+#endif
 }
 
 bool
 physaddr_t::init()
 {
 #ifdef LINUX
-    std::ostringstream oss;
-    std::string pagemap =
-        dynamic_cast<std::ostringstream &>(oss << "/proc/" << getpid() << "/pagemap")
-            .str();
+    // Some threads may not do much, so start out small.
+    constexpr int V2P_INITIAL_BITS = 9;
+    hashtable_init_ex(&v2p_, V2P_INITIAL_BITS, HASH_INTPTR, /*strdup=*/false,
+                      /*synch=*/false, nullptr, nullptr, nullptr);
+
+    // We avoid std::ostringstream to avoid malloc use for static linking.
+    constexpr int MAX_PAGEMAP_FNAME = 64;
+    char fname[MAX_PAGEMAP_FNAME];
+    dr_snprintf(fname, sizeof(fname), "/proc/%d/pagemap", getpid());
+    fname[MAX_PAGEMAP_FNAME - 1] = '\0';
```

Comment:
Is this correct? Don't we want to null-terminate after the "/proc/%d/pagemap",  instead of at `MAX_PAGEMAP_FNAME-1`?

Reply:
This is standard defensive programming with these string routines that do not always null-terminate (if the buffer fills up): in core/ this is NULL_TERMINATE_BUFFER.

Reply:
We have a style rule to always null-terminate a buffer after an snprintf or related operation.

Reply:
Actually I see drcachesim's utils.h has NULL_TERMINATE_BUFFER.  Changed to use it here.